### PR TITLE
gh-87729: specialize LOAD_SUPER_ATTR_METHOD

### LINKED ISSUE #87729: Use dedicated opcodes to speed up calls/attribute lookups with super() as receiver
BPO | [43563](https://bugs.python.org/issue43563)
--- | :---
Nosy | @gvanrossum, @rhettinger, @markshannon, @vladima, @isidentical
PRs | <li>python/cpython#24936</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-03-19.20:57:57.938>
labels = ['interpreter-core', '3.10']
title = 'Use dedicated opcodes to speed up calls/attribute lookups with super() as receiver'
updated_at = <Date 2021-04-27.18:23:03.781>
user = 'https://github.com/vladima'
```

bugs.python.org fields:
```python
activity = <Date 2021-04-27.18:23:03.781>
actor = 'v2m'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core']
creation = <Date 2021-03-19.20:57:57.938>
creator = 'v2m'
dependencies = []
files = []
hgrepos = []
issue_num = 43563
keywords = ['patch']
message_count = 7.0
messages = ['389114', '389156', '389168', '389337', '389340', '389344', '392116']
nosy_count = 5.0
nosy_names = ['gvanrossum', 'rhettinger', 'Mark.Shannon', 'v2m', 'BTaskaya']
pr_nums = ['24936']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue43563'
versions = ['Python 3.10']
```

</p></details>

### Linked PRs
* gh-103497
* gh-103809
* gh-104270
* gh-105125
* gh-105143

### PR DESCRIPTION
**Pull request by carljm**
fixes gh-87729

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -1584,6 +1600,19 @@ dummy_func(
             }
         }
 
+        inst(LOAD_SUPER_ATTR_METHOD, (unused/1, class_version/2, self_type_version/2, method/4, global_super, class, self -- res2, res)) {
+            DEOPT_IF(global_super != (PyObject *)&PySuper_Type, LOAD_SUPER_ATTR);
+            DEOPT_IF(!PyType_Check(class), LOAD_SUPER_ATTR);
+            DEOPT_IF(((PyTypeObject *)class)->tp_version_tag != class_version, LOAD_SUPER_ATTR);
+            PyTypeObject *self_type = Py_TYPE(self);
+            DEOPT_IF(self_type->tp_version_tag != self_type_version, LOAD_SUPER_ATTR);
```
Comment (markshannon) [inline Python/bytecodes.c:1592]: Given we expect a known self class and enclosing class, we also expect them to have a known relation.
So we expect `class` to be at a known index in `self.__class__`'s MRO.
We could (if we are seeking to shrink the cache) store the expected index into the MRO and check that (a single byte would be enough).

Probably wouldn't help performance though, as the additional indirection through the MRO wouldn't help.

**Review by markshannon — COMMENTED** [review]
There is some potential for simplifying `LOAD_SUPER_ATTR`, now that there is a specialization

File: Python/bytecodes.c (hunk 1 above)
Comment (carljm) [reply Python/bytecodes.c:1592]: We discussed this in person. It seems unlikely to improve perf, would be a bit harder to understand the correctness, and probably wouldn't even save much cache size due to alignment? So we can keep it in our back pocket as a future idea for cache shrinking, but we'll leave it as is for now.

**Review by markshannon — APPROVED** [review]
Looks good now.
Thanks.

File: Python/specialize.c [hunk 2]
```diff
@@ -505,6 +514,52 @@ specialize_module_load_attr(
...
+    if (global_super != (PyObject *)&PySuper_Type) {
+        SPECIALIZATION_FAIL(LOAD_SUPER_ATTR, SPEC_FAIL_SUPER_SHADOWED);
+        goto fail;
+    }
+    if (!PyType_Check(class)) {
+        SPECIALIZATION_FAIL(LOAD_SUPER_ATTR, SPEC_FAIL_SUPER_BAD_CLASS);
+        goto fail;
+    }
+    PyTypeObject *tp = (PyTypeObject *)class;
+    PyObject *res = _PySuper_LookupDescr(tp, self, name);
```
Comment (JelleZijlstra) [inline Python/specialize.c:536]: Is this leaked? I don't see where we DECREF it.

Comment (Fidget-Spinner) [reply Python/specialize.c:536]: Yeah I think it's leaked, _PySuper_LookupDescr incref's its result.

Comment (carljm) [reply Python/specialize.c:536]: For future reference, fix by @JelleZijlstra  in https://github.com/python/cpython/pull/103882

**Comment by tacaswell** [comment]
I think this is breaking c++ extensions as `class` is a reserved keyword in c++


```diff
diff --git a/Include/internal/pycore_code.h b/Include/internal/pycore_code.h
index 7d5d5e03de..f15328fd8e 100644
--- a/Include/internal/pycore_code.h
+++ b/Include/internal/pycore_code.h
@@ -226,7 +226,7 @@ extern int _PyLineTable_PreviousAddressRange(PyCodeAddressRange *range);

 /* Specialization functions */

-extern void _Py_Specialize_LoadSuperAttr(PyObject *global_super, PyObject *class, PyObject *self,
+extern void _Py_Specialize_LoadSuperAttr(PyObject *global_super, PyObject *klass, PyObject *self,
                                          _Py_CODEUNIT *instr, PyObject *name, int load_method);
 extern void _Py_Specialize_LoadAttr(PyObject *owner, _Py_CODEUNIT *instr,
                                     PyObject *name);
```

gets things compiling again, but I am not sure if this is the right fix (or you want to propagate the name change through the implementation as well).

**Comment by tacaswell** [comment]
Just this prototype and leave the rest?

**Comment by carljm** [comment]
All the other specialization functions include the parameter names in the prototype; it seems preferable to stay consistent. I don't mind changing `class` to `klass` throughout header and implementation; I can put up a PR for this.

**Pull request merged**